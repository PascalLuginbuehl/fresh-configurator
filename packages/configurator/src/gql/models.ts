export interface FlightControllerPartial {
  port: string;
  connection: { connecting: boolean; connected: boolean };
}

export interface ConnectionStatusMapping {
  parent: FlightControllerPartial;
}
