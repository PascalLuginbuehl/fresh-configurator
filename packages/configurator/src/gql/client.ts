import { ApolloClient, InMemoryCache } from "@apollo/client";
import {
  ports,
  isOpen,
  open,
  getAttitude,
  close,
  initialise,
  apiVersion,
  bytesRead,
  bytesWritten,
  packetErrors,
  getStatusExtended,
  getRcValues
} from "@fresh/msp";
import semver from "semver";
import {
  Resolvers,
  ConnectionStateQueryVariables,
  ConnectionStateDocument,
  ConnectionStateQuery,
  LogsQuery,
  LogsQueryVariables,
  LogsDocument
} from "./__generated__";
import config from "../config";
import { versionInfo } from "../util";

interface Context {
  client: ApolloClient<object>;
}

const cache = new InMemoryCache();

const selectedPort = cache.makeVar<string | null>(null);
// default baudrate
const selectedBaud = cache.makeVar<number>(115200);

const expertMode = cache.makeVar<boolean>(false);
const selectedTab = cache.makeVar<string | null>(null);

cache.policies.addTypePolicies({
  Configurator: {
    fields: {
      port: () => selectedPort(),
      baudRate: () => selectedBaud(),
      tab: () => selectedTab(),
      expertMode: () => expertMode()
    }
  }
});

const setConnectionState = (
  client: ApolloClient<object>,
  port: string,
  connecting: boolean,
  connected: boolean
): void =>
  client.writeQuery<ConnectionStateQuery, ConnectionStateQueryVariables>({
    query: ConnectionStateDocument,
    data: {
      __typename: "Query",
      device: {
        connection: {
          connected,
          connecting,
          __typename: "ConnectionStatus"
        },
        __typename: "FlightController"
      }
    },
    variables: {
      port
    }
  });

const log = (client: ApolloClient<object>, message: string): void => {
  const data = client.readQuery<LogsQuery, LogsQueryVariables>({
    query: LogsDocument
  });

  client.writeQuery<LogsQuery, LogsQueryVariables>({
    query: LogsDocument,
    data: {
      configurator: {
        logs: [
          ...(data?.configurator.logs ?? []).concat({
            time: new Date().toISOString(),
            message,
            __typename: "Log" as const
          })
        ],
        __typename: "Configurator"
      },
      __typename: "Query"
    }
  });
};

const resolvers: Resolvers = {
  Query: {
    ports: () => ports(),
    device: (_, { port }) => ({
      port,
      apiVersion: "",
      connection: {
        port,
        connecting: false,
        connected: false,
        __typename: "ConnectionStatus"
      },
      __typename: "FlightController"
    }),
    configurator: () => {
      const { os, version, chromeVersion } = versionInfo();
      return {
        port: selectedPort(),
        baudRate: selectedBaud(),
        tab: selectedTab(),
        expertMode: expertMode(),
        logs: [
          {
            time: new Date().toISOString(),
            message: `Running - OS: <strong>${os}</strong>, Chrome: <strong>${chromeVersion}</strong>, Configurator: <strong>${version}</strong>`,
            __typename: "Log"
          }
        ],
        __typename: "Configurator"
      };
    }
  },
  Mutation: {
    connect: async (_, { port, baudRate }, { client }: Context) => {
      if (isOpen(port)) {
        return true;
      }

      let closed = false;

      setConnectionState(client, port, true, false);

      await open(port, { baudRate }, () => {
        // on disconnect
        log(
          client,
          `Serial port <span class="message-positive">successfully</span> closed on ${port}`
        );
        setConnectionState(client, port, false, false);
        closed = true;
      });
      log(
        client,
        `Serial port <span class="message-positive">successfully</span> opened on ${port}`
      );

      try {
        await initialise(port);
        const version = apiVersion(port);
        log(client, `MultiWii API version: <strong>${version}</strong>`);

        if (semver.gte(version, config.apiVersionAccepted)) {
          setConnectionState(client, port, false, true);
          return true;
        }
        log(
          client,
          `MSP version not supported: <span class="message-negative">${version}</span>`
        );
      } catch (e) {
        console.log(e);
        log(
          client,
          `No configuration received within <span class="message-negative">5 seconds</span>, communication <span class="message-negative">failed</span>`
        );
      }

      // And only close the port if we are connecting
      if (!closed) {
        await close(port);
      }

      return false;
    },
    disconnect: async (_, { port }, { client }: Context) => {
      if (!isOpen(port)) {
        return true;
      }

      await close(port);
      setConnectionState(client, port, false, false);

      return true;
    },

    setTab: (_, { tabId }) => !!selectedTab(tabId),
    setConnectionSettings: (_, { port, baudRate }) => {
      selectedPort(port);
      if (typeof baudRate === "number") {
        selectedBaud(baudRate);
      }
      return true;
    },
    setExpertMode: (_, { enabled }) => !!expertMode(enabled),
    log: (_, { message }, { client }: Context) => {
      log(client, message);
      return true;
    }
  },

  FlightController: {
    attitude: ({ port }) =>
      getAttitude(port).then(values => ({ ...values, __typename: "Attitude" })),
    status: ({ port }) =>
      getStatusExtended(port).then(values => ({
        ...values,
        __typename: "Status"
      })),
    rc: ({ port }) => ({
      port,
      __typename: "RC"
    }),
    apiVersion: ({ port }) => apiVersion(port)
  },
  ConnectionStatus: {
    bytesRead: ({ port }) => bytesRead(port),
    bytesWritten: ({ port }) => bytesWritten(port),
    packetErrors: ({ port }) => packetErrors(port)
  },
  RC: {
    channels: ({ port }) => getRcValues(port)
  }
};

const client = new ApolloClient({
  cache,
  // generated resolvers are not compatible with apollo
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  resolvers: resolvers as any
});

export default client;
