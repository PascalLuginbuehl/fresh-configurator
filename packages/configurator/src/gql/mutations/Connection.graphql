mutation Connect($port: String!, $baudRate: Int!) {
  connect(port: $port, baudRate: $baudRate) @client
}

mutation Disconnect($port: String!) {
  disconnect(port: $port) @client
}
