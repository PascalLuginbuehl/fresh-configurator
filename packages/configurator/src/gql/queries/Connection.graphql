query AvailablePorts {
  ports
}

query ConnectionStats($connection: ID!) {
  connection(connectionId: $connection) {
    bytesWritten
    bytesRead
    packetErrors
  }
}

subscription OnConnectionChanged($connection: ID!) {
  onConnectionChanged(connectionId: $connection)
}
