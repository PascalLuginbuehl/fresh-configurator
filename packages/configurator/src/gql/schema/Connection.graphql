type Query {
  ports: [String!]!
}

type Mutation {
  connect(port: String!): Boolean
  disconnect(port: String!): Boolean
}
