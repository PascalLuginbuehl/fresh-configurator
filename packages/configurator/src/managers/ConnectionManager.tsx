import React, { useRef, useEffect, useReducer } from "react";
import semver from "semver";
import config from "../config";
import { UsbConnectIcon, UsbDisconnectIcon } from "../icons";
import useConnectionState from "../hooks/useConnectionState";
import useLogger from "../hooks/useLogger";
import {
  useConnectMutation,
  useConnectionSettingsQuery,
  useDisconnectMutation,
  useOnConnectionClosedSubscription,
} from "../gql/__generated__";
import BigButton from "../components/BigButton";

/**
 * Handle all aspects of tracking the app connection
 * to the device, as well as providing an interface
 * for the user to "connect" or "disconect"
 */
const ConnectionManager: React.FC = () => {
  const log = useLogger();

  const [, forceUpdate] = useReducer((x) => x + 1, 0);
  const attempt = useRef(0);
  const connectionAttempt = attempt.current;
  const isCurrentAttempt = (): boolean => attempt.current === connectionAttempt;

  const handleAborted = (): void => {
    attempt.current += 1;
    forceUpdate();
  };

  const { data: configuratorQuery } = useConnectionSettingsQuery();
  const { port, baudRate } = configuratorQuery?.configurator ?? {};

  const {
    connected,
    connecting,
    connection,
    setConnection,
    setConnecting,
  } = useConnectionState();

  const [disconnect] = useDisconnectMutation({
    variables: {
      connection: connection ?? "",
    },
    onCompleted: ({ close: connectionId }) => {
      if (!isCurrentAttempt()) {
        return;
      }
      log(
        `Serial port <span class="message-positive">successfully</span> closed for connectionId=${connectionId}`
      );
      setConnection(null);
    },
    onError: () => {
      if (!isCurrentAttempt()) {
        return;
      }
      setConnection(null);
    },
  });

  const [connect] = useConnectMutation({
    variables: {
      port: port ?? "",
      baudRate: baudRate ?? 0,
    },
    onCompleted: ({ connect: { id: connectionId, apiVersion } }) => {
      if (!connecting || !isCurrentAttempt()) {
        return;
      }

      log(
        `Serial port <span class="message-positive">successfully</span> opened, connectionId=${connectionId}`
      );
      log(`MultiWii API version: <strong>${apiVersion}</strong>`);
      if (semver.lt(apiVersion, config.apiVersionAccepted)) {
        log(
          `MSP version not supported: <span class="message-negative">${apiVersion}</span>`
        );
        disconnect({
          variables: {
            connection: connectionId,
          },
        });
      } else {
        setConnection(connectionId);
      }
      setConnecting(false);
    },
    onError: (e) => {
      if (!connecting || !isCurrentAttempt()) {
        return;
      }
      log(
        `Could not open connection (<span class="message-negative">${e.message}</span>), communication <span class="message-negative">failed</span>`
      );
      setConnecting(false);
    },
  });

  // Create a subscription to the current connection
  // and handle updating the app state when connection
  // is closed
  const { error: subscriptionError } = useOnConnectionClosedSubscription({
    variables: {
      connection: connection ?? "",
    },
    skip: !connection || !connected,
    onSubscriptionData: ({ subscriptionData }) => {
      const connectionId = subscriptionData.data?.onClosed;
      if (connectionId === connection) {
        setConnection(null);
        log(`Serial port closed unexpectedly for connectionId=${connectionId}`);
      }
    },
  });

  useEffect(() => {
    if (subscriptionError && connection) {
      setConnection(null);
      log(`Serial port closed unexpectedly for connectionId=${connection}`);
    }
  }, [connection, log, setConnection, subscriptionError]);

  const handleClicked = (): void => {
    if (!connected && !connecting) {
      log(`Opening connection on ${port}`);
      setConnecting(true);
      connect();
    } else {
      setConnecting(false);
      if (connection) {
        disconnect();
      }
      if (connecting && !connected) {
        log(`Connection attempt to ${port} aborted`);
        handleAborted();
      }
    }
  };

  let statusText = "Connecting";

  if (!connecting) {
    statusText = connected ? "Disconnect" : "Connect";
  }

  return (
    <BigButton
      active={connected}
      icon={
        !connected && !connecting ? <UsbConnectIcon /> : <UsbDisconnectIcon />
      }
      text={statusText}
      onClick={handleClicked}
    />
  );
};

export default ConnectionManager;
