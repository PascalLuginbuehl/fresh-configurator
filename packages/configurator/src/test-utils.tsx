import React from "react";
import { render as renderFunction, RenderResult } from "@testing-library/react";
import { ThemeProvider } from "./theme";

const AllTheProviders: React.FC = ({ children }) => (
  <ThemeProvider>{children}</ThemeProvider>
);

const customRender = (
  ui: Parameters<typeof renderFunction>[0],
  options?: Parameters<typeof renderFunction>[1]
): RenderResult => renderFunction(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
