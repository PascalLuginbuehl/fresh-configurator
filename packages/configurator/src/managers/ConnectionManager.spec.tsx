import React from "react";
import { Resolvers } from "apollo-client";
import { MockedProvider, MockedResponse } from "@apollo/react-testing";
import { InMemoryCache } from "apollo-cache-inmemory";
import { render, waitFor, fireEvent } from "../test-utils";
import { ConnectDocument } from "../gql/mutations/Connection.graphql";
import { Connection } from "../gql/__generated__";
import ConnectionManager from "./ConnectionManager";

let logs: string[] = [];

const configuratorState = (
  state: {
    connecting?: boolean;
    connection?: string | null;
    port?: string;
    baudRate?: number;
  },
  onConnectionSet = (connection: string) => connection,
  onConnectingSet = (value: boolean) => value
): Resolvers => ({
  Query: {
    configurator: () => ({
      __typename: "Configurator",

      port: "",
      baudRate: 115200,
      connecting: false,
      connection: null,

      ...state,
    }),
  },
  Mutation: {
    setConnection: (_, { connection }) => {
      onConnectionSet(connection);
      return null;
    },
    setConnecting: (_, { value }) => {
      onConnectingSet(value);
      return null;
    },
    log: (_, { message }) => {
      logs.push(message);
      return null;
    },
  },
});

const connectMock = (
  port: string,
  baudRate: number,
  connectionId: string,
  apiVersion: string,
  onConnect: () => void
): MockedResponse => ({
  request: {
    query: ConnectDocument,
    variables: {
      port,
      baudRate,
    },
  },
  result: () => {
    onConnect();

    return {
      data: {
        connect: {
          id: connectionId,
          apiVersion,
          __typename: "Connection",
        } as Connection,
      },
    };
  },
});

beforeEach(() => {
  logs = [];
});

describe("ConnectionManager", () => {
  xit("should connect with the connection settings and set the new connection id when connect clicked", async () => {
    const connect = jest.fn();
    const setConnecting = jest.fn();
    const setConnection = jest.fn();
    const mockConnectionId = "23455";

    const cache = new InMemoryCache({
      resultCaching: false,
    });

    const { getByTestId, asFragment, getByText, rerender } = render(
      <MockedProvider
        cache={cache}
        resolvers={configuratorState(
          {
            connecting: false,
            connection: null,
            port: "/dev/something",
            baudRate: 115200,
          },
          setConnection,
          setConnecting
        )}
        mocks={[
          connectMock(
            "/dev/something",
            115200,
            mockConnectionId,
            "1.40.1",
            connect
          ),
        ]}
      >
        <ConnectionManager />
      </MockedProvider>
    );

    await waitFor(() => expect(getByTestId("connect")).toBeEnabled());

    // expect the button to look correct
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(getByTestId("connect"));

    await waitFor(() => expect(setConnecting).toHaveBeenCalledWith(true));

    cache.reset();
    rerender(
      <MockedProvider
        cache={cache}
        resolvers={configuratorState(
          {
            connecting: true,
            connection: null,
            port: "/dev/something",
            baudRate: 115200,
          },
          setConnection,
          setConnecting
        )}
        mocks={[
          connectMock(
            "/dev/something",
            115200,
            mockConnectionId,
            "1.40.1",
            connect
          ),
        ]}
      >
        <ConnectionManager />
      </MockedProvider>
    );
    await waitFor(() => expect(getByText("Connecting")).toBeVisible());
    await waitFor(() => expect(connect).toHaveBeenCalled());
    await waitFor(() =>
      expect(setConnection).toHaveBeenCalledWith(mockConnectionId)
    );

    expect(logs).toMatchSnapshot();
  });
});
