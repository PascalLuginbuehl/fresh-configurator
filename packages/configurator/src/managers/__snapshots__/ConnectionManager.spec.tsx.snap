// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectionManager should connect with the connection settings and set the new connection id when connect clicked 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 70px;
  width: 70px;
  color: #fff;
  font-size: 12px;
  font-weight: normal;
  text-shadow: 0px 1px rgba(0,0,0,0.25);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #ffbb00;
  border: 1px solid #dba718;
  height: 50px;
  width: 50px;
  border-radius: 100px;
  padding-top: 0px;
  padding-right: 0px;
  padding-bottom: 0px;
  padding-left: 0px;
  cursor: pointer;
}

.c1 svg {
  width: 44px;
  height: 44px;
}

.c1:hover {
  background-color: #ffcc3f;
}

<div
    class="c0"
  >
    <button
      class="c1"
      data-testid="connect"
      type="button"
    >
      <svg
        data-name="usb-connect"
      />
    </button>
    <div>
      Connect
    </div>
  </div>
</DocumentFragment>
`;
